{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "63097c05",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.Data import CodonTable\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "081ea136",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "800\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>site</th>\n",
       "      <th>gene</th>\n",
       "      <th>direction</th>\n",
       "      <th>snp</th>\n",
       "      <th>dimer</th>\n",
       "      <th>apobec</th>\n",
       "      <th>aa_position</th>\n",
       "      <th>parent</th>\n",
       "      <th>parent_codon</th>\n",
       "      <th>parent_aa</th>\n",
       "      <th>child</th>\n",
       "      <th>child_codon</th>\n",
       "      <th>child_aa</th>\n",
       "      <th>mutation_category</th>\n",
       "      <th>score</th>\n",
       "      <th>prediction</th>\n",
       "      <th>homoplasy</th>\n",
       "      <th>occurrence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3111</td>\n",
       "      <td>OPG003_CDS_176</td>\n",
       "      <td>reverse</td>\n",
       "      <td>G-&gt;A</td>\n",
       "      <td>GA</td>\n",
       "      <td>True</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Node1</td>\n",
       "      <td>TCC</td>\n",
       "      <td>S</td>\n",
       "      <td>Node2</td>\n",
       "      <td>TTC</td>\n",
       "      <td>F</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>155.0</td>\n",
       "      <td>radical</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>39139</td>\n",
       "      <td>OPG057_CDS_142</td>\n",
       "      <td>reverse</td>\n",
       "      <td>C-&gt;T</td>\n",
       "      <td>TC</td>\n",
       "      <td>True</td>\n",
       "      <td>3.0</td>\n",
       "      <td>Node1</td>\n",
       "      <td>GCG</td>\n",
       "      <td>A</td>\n",
       "      <td>Node2</td>\n",
       "      <td>GCA</td>\n",
       "      <td>A</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>73239</td>\n",
       "      <td>OPG093_CDS_107</td>\n",
       "      <td>forward</td>\n",
       "      <td>G-&gt;A</td>\n",
       "      <td>GA</td>\n",
       "      <td>True</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Node1</td>\n",
       "      <td>AGA</td>\n",
       "      <td>R</td>\n",
       "      <td>Node2</td>\n",
       "      <td>AAA</td>\n",
       "      <td>K</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>26.0</td>\n",
       "      <td>conservative</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>74205</td>\n",
       "      <td>OPG094_CDS_106</td>\n",
       "      <td>forward</td>\n",
       "      <td>G-&gt;A</td>\n",
       "      <td>GA</td>\n",
       "      <td>True</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Node1</td>\n",
       "      <td>GAA</td>\n",
       "      <td>E</td>\n",
       "      <td>Node2</td>\n",
       "      <td>AAA</td>\n",
       "      <td>K</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>56.0</td>\n",
       "      <td>moderately conservative</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>77383</td>\n",
       "      <td>OPG098_CDS_102</td>\n",
       "      <td>forward</td>\n",
       "      <td>G-&gt;A</td>\n",
       "      <td>GA</td>\n",
       "      <td>True</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Node1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>*</td>\n",
       "      <td>Node2</td>\n",
       "      <td>TAA</td>\n",
       "      <td>*</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    site            gene direction   snp dimer  apobec  aa_position parent  \\\n",
       "0   3111  OPG003_CDS_176   reverse  G->A    GA    True          2.0  Node1   \n",
       "1  39139  OPG057_CDS_142   reverse  C->T    TC    True          3.0  Node1   \n",
       "2  73239  OPG093_CDS_107   forward  G->A    GA    True          2.0  Node1   \n",
       "3  74205  OPG094_CDS_106   forward  G->A    GA    True          1.0  Node1   \n",
       "4  77383  OPG098_CDS_102   forward  G->A    GA    True          2.0  Node1   \n",
       "\n",
       "  parent_codon parent_aa  child child_codon child_aa mutation_category  score  \\\n",
       "0          TCC         S  Node2         TTC        F     nonsynonymous  155.0   \n",
       "1          GCG         A  Node2         GCA        A        synonymous    NaN   \n",
       "2          AGA         R  Node2         AAA        K     nonsynonymous   26.0   \n",
       "3          GAA         E  Node2         AAA        K     nonsynonymous   56.0   \n",
       "4          TGA         *  Node2         TAA        *        synonymous    NaN   \n",
       "\n",
       "                prediction  homoplasy  occurrence  \n",
       "0                  radical      False           1  \n",
       "1                      NaN      False           1  \n",
       "2             conservative      False           1  \n",
       "3  moderately conservative      False           1  \n",
       "4                      NaN      False           1  "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "article_pos_file = \"data/B.1_2022-08-22.og.aln.pruned.tree.amino_acid.reconstruction.csv\"\n",
    "article_df = pd.read_csv(article_pos_file, sep=',')\n",
    "print(len(article_df))\n",
    "article_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "c616a053",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['site', 'gene', 'direction', 'snp', 'dimer', 'apobec', 'aa_position',\n",
       "       'parent', 'parent_codon', 'parent_aa', 'child', 'child_codon',\n",
       "       'child_aa', 'mutation_category', 'score', 'prediction', 'homoplasy',\n",
       "       'occurrence'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "article_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "e93d2114",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>site</th>\n",
       "      <th>gene</th>\n",
       "      <th>direction</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>dimer</th>\n",
       "      <th>apobec</th>\n",
       "      <th>aa_position</th>\n",
       "      <th>parent</th>\n",
       "      <th>parent_codon</th>\n",
       "      <th>parent_aa</th>\n",
       "      <th>child</th>\n",
       "      <th>child_codon</th>\n",
       "      <th>child_aa</th>\n",
       "      <th>mutation_category</th>\n",
       "      <th>score</th>\n",
       "      <th>prediction</th>\n",
       "      <th>homoplasy</th>\n",
       "      <th>occurrence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3111</td>\n",
       "      <td>OPG003_CDS_176</td>\n",
       "      <td>reverse</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>GA</td>\n",
       "      <td>True</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Node1</td>\n",
       "      <td>TCC</td>\n",
       "      <td>S</td>\n",
       "      <td>Node2</td>\n",
       "      <td>TTC</td>\n",
       "      <td>F</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>155.0</td>\n",
       "      <td>radical</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>39139</td>\n",
       "      <td>OPG057_CDS_142</td>\n",
       "      <td>reverse</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>TC</td>\n",
       "      <td>True</td>\n",
       "      <td>3.0</td>\n",
       "      <td>Node1</td>\n",
       "      <td>GCG</td>\n",
       "      <td>A</td>\n",
       "      <td>Node2</td>\n",
       "      <td>GCA</td>\n",
       "      <td>A</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>73239</td>\n",
       "      <td>OPG093_CDS_107</td>\n",
       "      <td>forward</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>GA</td>\n",
       "      <td>True</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Node1</td>\n",
       "      <td>AGA</td>\n",
       "      <td>R</td>\n",
       "      <td>Node2</td>\n",
       "      <td>AAA</td>\n",
       "      <td>K</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>26.0</td>\n",
       "      <td>conservative</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>74205</td>\n",
       "      <td>OPG094_CDS_106</td>\n",
       "      <td>forward</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>GA</td>\n",
       "      <td>True</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Node1</td>\n",
       "      <td>GAA</td>\n",
       "      <td>E</td>\n",
       "      <td>Node2</td>\n",
       "      <td>AAA</td>\n",
       "      <td>K</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>56.0</td>\n",
       "      <td>moderately conservative</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>77383</td>\n",
       "      <td>OPG098_CDS_102</td>\n",
       "      <td>forward</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>GA</td>\n",
       "      <td>True</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Node1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>*</td>\n",
       "      <td>Node2</td>\n",
       "      <td>TAA</td>\n",
       "      <td>*</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    site            gene direction REF ALT dimer  apobec  aa_position parent  \\\n",
       "0   3111  OPG003_CDS_176   reverse   G   A    GA    True          2.0  Node1   \n",
       "1  39139  OPG057_CDS_142   reverse   C   T    TC    True          3.0  Node1   \n",
       "2  73239  OPG093_CDS_107   forward   G   A    GA    True          2.0  Node1   \n",
       "3  74205  OPG094_CDS_106   forward   G   A    GA    True          1.0  Node1   \n",
       "4  77383  OPG098_CDS_102   forward   G   A    GA    True          2.0  Node1   \n",
       "\n",
       "  parent_codon parent_aa  child child_codon child_aa mutation_category  score  \\\n",
       "0          TCC         S  Node2         TTC        F     nonsynonymous  155.0   \n",
       "1          GCG         A  Node2         GCA        A        synonymous    NaN   \n",
       "2          AGA         R  Node2         AAA        K     nonsynonymous   26.0   \n",
       "3          GAA         E  Node2         AAA        K     nonsynonymous   56.0   \n",
       "4          TGA         *  Node2         TAA        *        synonymous    NaN   \n",
       "\n",
       "                prediction  homoplasy  occurrence  \n",
       "0                  radical      False           1  \n",
       "1                      NaN      False           1  \n",
       "2             conservative      False           1  \n",
       "3  moderately conservative      False           1  \n",
       "4                      NaN      False           1  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "article_df['REF'] = article_df.apply(lambda row: row['snp'].split('->')[0], axis = 1)\n",
    "article_df['ALT'] = article_df.apply(lambda row: row['snp'].split('->')[1], axis = 1)\n",
    "article_df = article_df[['site', 'gene', 'direction', 'REF', \"ALT\", 'dimer', 'apobec', 'aa_position',\n",
    "       'parent', 'parent_codon', 'parent_aa', 'child', 'child_codon',\n",
    "       'child_aa', 'mutation_category', 'score', 'prediction', 'homoplasy',\n",
    "       'occurrence']]\n",
    "article_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d5fbe87a",
   "metadata": {},
   "source": [
    "Searching for amino acids in positions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "b7265889",
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_genome(fasta_file):\n",
    "    \"\"\"Load the genome from a FASTA file.\"\"\"\n",
    "    genome = SeqIO.to_dict(SeqIO.parse(fasta_file, \"fasta\"))\n",
    "    new_genome = genome[\"NC_063383.1\"].seq\n",
    "    return new_genome\n",
    "\n",
    "def load_gtf(gtf_file):\n",
    "    \"\"\"Load the GTF file into a pandas DataFrame.\"\"\"\n",
    "    gtf_columns = [\"seqname\", \"source\", \"feature\", \"start\", \"end\", \"score\", \"strand\", \"frame\", \"attribute\"]\n",
    "    gtf = pd.read_csv(gtf_file, sep=\"\\t\", comment='#', header=None, names=gtf_columns)\n",
    "    return gtf\n",
    "\n",
    "def find_node_genome(genome, parent, table=article_df):\n",
    "    parent_filter_df = table[table['child']==parent]\n",
    "    parent_filter_df = parent_filter_df.reset_index(drop=True)\n",
    "    new_parent = parent_filter_df.loc[0, \"parent\"]\n",
    "    sites_list = parent_filter_df.site.tolist()\n",
    "    \n",
    "    if new_parent == \"Node1\":\n",
    "        new_genome = list(genome)\n",
    "        for row in range(0, len(parent_filter_df)):\n",
    "            position = parent_filter_df.loc[row, \"site\"]\n",
    "            alt_nucl = parent_filter_df.loc[row, \"ALT\"]\n",
    "            new_genome[position-1] = alt_nucl\n",
    "        return Seq(''.join(new_genome))\n",
    "    else:\n",
    "        new_genome = list(genome)\n",
    "        for row in range(0, len(parent_filter_df)):\n",
    "            position = parent_filter_df.loc[row, \"site\"]\n",
    "            alt_nucl = parent_filter_df.loc[row, \"ALT\"]\n",
    "            new_genome[position-1] = alt_nucl\n",
    "        return find_node_genome(new_genome, new_parent)\n",
    "\n",
    "    \n",
    "def get_cds(gtf, chromosome, position):\n",
    "    \"\"\"Find the CDS where the mutation is located.\"\"\"\n",
    "    cds_rows = gtf[(gtf['seqname'] == chromosome) & \n",
    "                   (gtf['feature'] == 'CDS') & \n",
    "                   (gtf['start'] <= position) & \n",
    "                   (gtf['end'] >= position)]\n",
    "    return cds_rows\n",
    "\n",
    "def translate_codon(codon):\n",
    "    \"\"\"Translate a codon to its corresponding amino acid.\"\"\"\n",
    "    return str(Seq(codon).translate(table=CodonTable.unambiguous_dna_by_id[11]))\n",
    "\n",
    "def detect_mutation_impact(genome, gtf, chromosome, position, ref, alt, gene_name):\n",
    "    \"\"\"Detect the amino acid change caused by a mutation.\"\"\"\n",
    "    cds = get_cds(gtf, chromosome, position)\n",
    "    \n",
    "    if cds.empty:\n",
    "        ref_nucl = genome[position-1]\n",
    "        return '-', '-', '-', '-', ref_nucl, '-'\n",
    "    \n",
    "    row_id = 0\n",
    "    for row in range(0, len(cds)):\n",
    "        info = cds.iloc[row]['attribute'].split(';')\n",
    "        for x in info:\n",
    "            if x.startswith(\"gene=\"):\n",
    "                gene = x.split(\"=\")[1]\n",
    "                if gene == gene_name:\n",
    "                    row_id = row\n",
    "    \n",
    "    strand = cds.iloc[row_id]['strand']\n",
    "    gene_start = cds.iloc[row_id]['start']\n",
    "    \n",
    "    # Calculate the codon position\n",
    "    codon_start = gene_start + ((position - gene_start) // 3) * 3\n",
    "    \n",
    "    # Extract the codon\n",
    "    codon_seq = genome[codon_start-1:codon_start+2]\n",
    "    if strand == '-':\n",
    "        codon_seq = codon_seq.reverse_complement()\n",
    "        alt = str(Seq(alt).reverse_complement())\n",
    "    \n",
    "    # Translate the original codon\n",
    "    original_amino_acid = translate_codon(codon_seq)\n",
    "    \n",
    "    # Create the mutated codon\n",
    "    mutation_index = (position - codon_start) % 3\n",
    "    if strand == '-':\n",
    "        if mutation_index == 0:\n",
    "            mutation_index = 2\n",
    "        elif mutation_index == 2:\n",
    "            mutation_index = 0\n",
    "\n",
    "    mutated_codon = list(codon_seq)\n",
    "    mutated_codon[mutation_index] = alt\n",
    "    mutated_codon = Seq(''.join(mutated_codon))\n",
    "    \n",
    "    # Translate the mutated codon\n",
    "    mutated_amino_acid = translate_codon(mutated_codon)\n",
    "    \n",
    "    ref_nucl = genome[position-1]\n",
    "    pos = mutation_index+1\n",
    "    \n",
    "    return original_amino_acid, mutated_amino_acid, codon_seq, mutated_codon, ref_nucl, pos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "0addc80b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "800\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>site</th>\n",
       "      <th>gene</th>\n",
       "      <th>direction</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>dimer</th>\n",
       "      <th>apobec</th>\n",
       "      <th>aa_position</th>\n",
       "      <th>parent</th>\n",
       "      <th>parent_codon</th>\n",
       "      <th>...</th>\n",
       "      <th>prediction</th>\n",
       "      <th>homoplasy</th>\n",
       "      <th>occurrence</th>\n",
       "      <th>REF_v2</th>\n",
       "      <th>aa_position_FIXED</th>\n",
       "      <th>parent_codon_FIXED</th>\n",
       "      <th>parent_aa_FIXED</th>\n",
       "      <th>child_codon_FIXED</th>\n",
       "      <th>child_aa_FIXED</th>\n",
       "      <th>mutation_category_FIXED</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3111</td>\n",
       "      <td>OPG003_CDS_176</td>\n",
       "      <td>reverse</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>GA</td>\n",
       "      <td>True</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Node1</td>\n",
       "      <td>TCC</td>\n",
       "      <td>...</td>\n",
       "      <td>radical</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>G</td>\n",
       "      <td>3</td>\n",
       "      <td>ATC</td>\n",
       "      <td>I</td>\n",
       "      <td>ATT</td>\n",
       "      <td>I</td>\n",
       "      <td>synonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>39139</td>\n",
       "      <td>OPG057_CDS_142</td>\n",
       "      <td>reverse</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>TC</td>\n",
       "      <td>True</td>\n",
       "      <td>3.0</td>\n",
       "      <td>Node1</td>\n",
       "      <td>GCG</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>GAG</td>\n",
       "      <td>E</td>\n",
       "      <td>AAG</td>\n",
       "      <td>K</td>\n",
       "      <td>nonsynonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>73239</td>\n",
       "      <td>OPG093_CDS_107</td>\n",
       "      <td>forward</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>GA</td>\n",
       "      <td>True</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Node1</td>\n",
       "      <td>AGA</td>\n",
       "      <td>...</td>\n",
       "      <td>conservative</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>G</td>\n",
       "      <td>1</td>\n",
       "      <td>GAT</td>\n",
       "      <td>D</td>\n",
       "      <td>AAT</td>\n",
       "      <td>N</td>\n",
       "      <td>nonsynonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>74205</td>\n",
       "      <td>OPG094_CDS_106</td>\n",
       "      <td>forward</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>GA</td>\n",
       "      <td>True</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Node1</td>\n",
       "      <td>GAA</td>\n",
       "      <td>...</td>\n",
       "      <td>moderately conservative</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>G</td>\n",
       "      <td>3</td>\n",
       "      <td>ATG</td>\n",
       "      <td>M</td>\n",
       "      <td>ATA</td>\n",
       "      <td>I</td>\n",
       "      <td>nonsynonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>77383</td>\n",
       "      <td>OPG098_CDS_102</td>\n",
       "      <td>forward</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>GA</td>\n",
       "      <td>True</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Node1</td>\n",
       "      <td>TGA</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>G</td>\n",
       "      <td>1</td>\n",
       "      <td>GAG</td>\n",
       "      <td>E</td>\n",
       "      <td>AAG</td>\n",
       "      <td>K</td>\n",
       "      <td>nonsynonymous</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    site            gene direction REF ALT dimer  apobec  aa_position parent  \\\n",
       "0   3111  OPG003_CDS_176   reverse   G   A    GA    True          2.0  Node1   \n",
       "1  39139  OPG057_CDS_142   reverse   C   T    TC    True          3.0  Node1   \n",
       "2  73239  OPG093_CDS_107   forward   G   A    GA    True          2.0  Node1   \n",
       "3  74205  OPG094_CDS_106   forward   G   A    GA    True          1.0  Node1   \n",
       "4  77383  OPG098_CDS_102   forward   G   A    GA    True          2.0  Node1   \n",
       "\n",
       "  parent_codon  ...               prediction homoplasy occurrence REF_v2  \\\n",
       "0          TCC  ...                  radical     False          1      G   \n",
       "1          GCG  ...                      NaN     False          1      C   \n",
       "2          AGA  ...             conservative     False          1      G   \n",
       "3          GAA  ...  moderately conservative     False          1      G   \n",
       "4          TGA  ...                      NaN     False          1      G   \n",
       "\n",
       "  aa_position_FIXED  parent_codon_FIXED parent_aa_FIXED  child_codon_FIXED  \\\n",
       "0                 3                 ATC               I                ATT   \n",
       "1                 1                 GAG               E                AAG   \n",
       "2                 1                 GAT               D                AAT   \n",
       "3                 3                 ATG               M                ATA   \n",
       "4                 1                 GAG               E                AAG   \n",
       "\n",
       "   child_aa_FIXED mutation_category_FIXED  \n",
       "0               I              synonymous  \n",
       "1               K           nonsynonymous  \n",
       "2               N           nonsynonymous  \n",
       "3               I           nonsynonymous  \n",
       "4               K           nonsynonymous  \n",
       "\n",
       "[5 rows x 26 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genome = load_genome(\"data/MPOX_genome/NC_063383.1.fasta\")\n",
    "gtf = load_gtf(\"data/MPOX_genome/GCF_014621545.1_ASM1462154v1_genomic.220824.gff\")\n",
    "chromosome = \"NC_063383.1\"\n",
    "\n",
    "original_aa_list = []\n",
    "mutated_aa_list = []\n",
    "original_codon_list = []\n",
    "mutated_codon_list = []\n",
    "ref_nucl_list = []\n",
    "nucl_pos_list = []\n",
    "mutation_category = []\n",
    "\n",
    "\n",
    "for row in range(0, len(article_df)):\n",
    "    position = article_df.loc[row, \"site\"] \n",
    "    ref_nucl = article_df.loc[row, \"REF\"]\n",
    "    alt_nucl = article_df.loc[row, \"ALT\"]\n",
    "    gene = article_df.loc[row, \"gene\"]\n",
    "    if gene is not np.nan:\n",
    "        gene = gene.split('_')[0]\n",
    "    node = article_df.loc[row, \"parent\"]\n",
    "    \n",
    "    if node == \"Node1\":\n",
    "        original_aa, mutated_aa, original_codon, mutated_codon, ref, nucl_pos = detect_mutation_impact(genome, gtf, chromosome, position, ref_nucl, alt_nucl, gene)\n",
    "    else:  \n",
    "        new_genome = find_node_genome(genome, node)\n",
    "        original_aa, mutated_aa, original_codon, mutated_codon, ref, nucl_pos = detect_mutation_impact(new_genome, gtf, chromosome, position, ref_nucl, alt_nucl, gene)\n",
    "    \n",
    "    # CDS empty\n",
    "    if original_aa == \"-\":\n",
    "        mutation_category.append(\"intergenic\")\n",
    "        original_aa_list.append('-')\n",
    "        mutated_aa_list.append('-')\n",
    "        original_codon_list.append('-')\n",
    "        mutated_codon_list.append('-')\n",
    "        nucl_pos_list.append('-')\n",
    "        ref_nucl_list.append(ref)\n",
    "    else:\n",
    "        original_aa_list.append(original_aa)\n",
    "        mutated_aa_list.append(mutated_aa)\n",
    "        original_codon_list.append(str(original_codon))\n",
    "        mutated_codon_list.append(str(mutated_codon))\n",
    "        ref_nucl_list.append(ref)\n",
    "        nucl_pos_list.append(nucl_pos)\n",
    "\n",
    "        if mutated_aa == original_aa:\n",
    "            mutation_category.append(\"synonymous\")\n",
    "        elif mutated_aa == \"*\":\n",
    "            mutation_category.append(\"nonsense\")\n",
    "        else:\n",
    "            mutation_category.append(\"nonsynonymous\")\n",
    "            \n",
    "article_df[\"REF_v2\"] = ref_nucl_list\n",
    "article_df[\"aa_position_FIXED\"] = nucl_pos_list\n",
    "article_df[\"parent_codon_FIXED\"] = original_codon_list\n",
    "article_df[\"parent_aa_FIXED\"] = original_aa_list\n",
    "article_df[\"child_codon_FIXED\"] = mutated_codon_list\n",
    "article_df[\"child_aa_FIXED\"] = mutated_aa_list\n",
    "article_df[\"mutation_category_FIXED\"] = mutation_category\n",
    "print(len(article_df))\n",
    "article_df.head()\n",
    "article_df.to_csv(\"data/B1_data_fixed.csv\", sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "b2ce08e3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "does not match to reference:  2\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>site</th>\n",
       "      <th>gene</th>\n",
       "      <th>direction</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>dimer</th>\n",
       "      <th>apobec</th>\n",
       "      <th>aa_position</th>\n",
       "      <th>parent</th>\n",
       "      <th>parent_codon</th>\n",
       "      <th>...</th>\n",
       "      <th>homoplasy</th>\n",
       "      <th>occurrence</th>\n",
       "      <th>REF_v2</th>\n",
       "      <th>aa_position_FIXED</th>\n",
       "      <th>parent_codon_FIXED</th>\n",
       "      <th>parent_aa_FIXED</th>\n",
       "      <th>child_codon_FIXED</th>\n",
       "      <th>child_aa_FIXED</th>\n",
       "      <th>mutation_category_FIXED</th>\n",
       "      <th>new_col</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>268</th>\n",
       "      <td>168120</td>\n",
       "      <td>OPG193_CDS_17</td>\n",
       "      <td>forward</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Node386</td>\n",
       "      <td>TTT</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>CTT</td>\n",
       "      <td>L</td>\n",
       "      <td>CTT</td>\n",
       "      <td>L</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>423</th>\n",
       "      <td>81977</td>\n",
       "      <td>OPG105_CDS_95</td>\n",
       "      <td>forward</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>GA</td>\n",
       "      <td>True</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Node561</td>\n",
       "      <td>GAC</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>3</td>\n",
       "      <td>CAA</td>\n",
       "      <td>Q</td>\n",
       "      <td>CAA</td>\n",
       "      <td>Q</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       site           gene direction REF ALT dimer  apobec  aa_position  \\\n",
       "268  168120  OPG193_CDS_17   forward   T   C   NaN   False          2.0   \n",
       "423   81977  OPG105_CDS_95   forward   G   A    GA    True          1.0   \n",
       "\n",
       "      parent parent_codon  ... homoplasy occurrence REF_v2 aa_position_FIXED  \\\n",
       "268  Node386          TTT  ...     False          1      C                 1   \n",
       "423  Node561          GAC  ...     False          1      A                 3   \n",
       "\n",
       "    parent_codon_FIXED  parent_aa_FIXED child_codon_FIXED  child_aa_FIXED  \\\n",
       "268                CTT                L               CTT               L   \n",
       "423                CAA                Q               CAA               Q   \n",
       "\n",
       "     mutation_category_FIXED new_col  \n",
       "268               synonymous       -  \n",
       "423               synonymous       -  \n",
       "\n",
       "[2 rows x 27 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_col = []\n",
    "for row in range(0, len(article_df)):\n",
    "    if article_df.loc[row, \"REF\"] != article_df.loc[row, \"REF_v2\"]:\n",
    "        new_col.append('-')\n",
    "    else:\n",
    "        new_col.append(\"+\")\n",
    "article_df[\"new_col\"] = new_col\n",
    "print('does not match to reference: ', len(article_df[article_df['new_col']=='-']))\n",
    "article_df[article_df['new_col']=='-']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "2628f884",
   "metadata": {},
   "outputs": [],
   "source": [
    "#filter only C->T and G->A mutations\n",
    "df1_new1 = article_df[(article_df['REF_v2']==\"G\") & (article_df['ALT']==\"A\")]\n",
    "df1_new2 = article_df[(article_df['REF_v2']==\"C\") & (article_df['ALT']==\"T\")]\n",
    "df1_new = pd.concat([df1_new1, df1_new2])\n",
    "df1_new = df1_new.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "97e0b6e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "#filter TC and GA motifs\n",
    "def load_genome(fasta_file):\n",
    "    \"\"\"Load the genome from a FASTA file.\"\"\"\n",
    "    genome = SeqIO.to_dict(SeqIO.parse(fasta_file, \"fasta\"))\n",
    "    new_genome = genome[\"NC_063383.1\"].seq\n",
    "    return str(new_genome)\n",
    "\n",
    "def find_node_genome(genome, parent, table=article_df):\n",
    "    parent_filter_df = table[table['child']==parent]\n",
    "    parent_filter_df = parent_filter_df.reset_index(drop=True)\n",
    "    new_parent = parent_filter_df.loc[0, \"parent\"]\n",
    "    sites_list = parent_filter_df.site.tolist()\n",
    "    \n",
    "    if new_parent == \"Node1\":\n",
    "        new_genome = list(genome)\n",
    "        for row in range(0, len(parent_filter_df)):\n",
    "            position = parent_filter_df.loc[row, \"site\"]\n",
    "            alt_nucl = parent_filter_df.loc[row, \"ALT\"]\n",
    "            new_genome[position-1] = alt_nucl\n",
    "        return Seq(''.join(new_genome))\n",
    "    else:\n",
    "        new_genome = list(genome)\n",
    "        for row in range(0, len(parent_filter_df)):\n",
    "            position = parent_filter_df.loc[row, \"site\"]\n",
    "            alt_nucl = parent_filter_df.loc[row, \"ALT\"]\n",
    "            new_genome[position-1] = alt_nucl\n",
    "        return find_node_genome(new_genome, new_parent)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "1897770b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "641\n"
     ]
    }
   ],
   "source": [
    "genome = load_genome(\"data/MPOX_genome/NC_063383.1.fasta\")\n",
    "motifs = []\n",
    "for row in range(0, len(df1_new)):\n",
    "    pos = df1_new.loc[row, 'site']\n",
    "    ref = df1_new.loc[row, 'REF']\n",
    "    node = df1_new.loc[row, 'parent']\n",
    "    if node != \"Node1\":\n",
    "        new_genome = find_node_genome(genome, node)\n",
    "        if ref == \"C\":\n",
    "            motifs.append(new_genome[pos-2:pos])\n",
    "        else:\n",
    "            motifs.append(new_genome[pos-1:pos+1])\n",
    "    else:\n",
    "        if ref == \"C\":\n",
    "            motifs.append(genome[pos-2:pos])\n",
    "        else:\n",
    "            motifs.append(genome[pos-1:pos+1])\n",
    "\n",
    "df1_new['motif'] = motifs\n",
    "\n",
    "df1_new1 = df1_new[df1_new['motif'] == 'GA']\n",
    "df1_new2 = df1_new[df1_new['motif'] == 'TC']\n",
    "df1_new = pd.concat([df1_new1, df1_new2])\n",
    "print(len(df1_new))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "5565595c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "631"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#filter Node1\n",
    "df1_new = df1_new[df1_new['parent'] != 'Node1']\n",
    "df1_new = df1_new.reset_index(drop=True)\n",
    "len(df1_new)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c9a3cc7a",
   "metadata": {},
   "source": [
    "Searching for APOBEC3 targets in MPOX genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "9f224c0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "\n",
    "## genome file\n",
    "genome_file = open(\"data/MPOX_genome/NC_063383.1.fasta\")\n",
    "genome_dict = SeqIO.to_dict(SeqIO.parse(genome_file, \"fasta\"))\n",
    "\n",
    "## find coordinates of potential targets with motif in sample\n",
    "## output: list with potential targets in the genome for sample\n",
    "def find_potential_targets(signature='TC', genome=str(genome_dict['NC_063383.1'].seq)):\n",
    "    signature_complement = str(Seq(signature).reverse_complement())\n",
    "    \n",
    "    #genome start with 0, write position for indexing from 1 \n",
    "    targets_coordinates_TC = [i.start()+2 for i in re.finditer(signature, genome, flags=0)]\n",
    "    targets_coordinates_GA = [i.start()+1 for i in re.finditer(signature_complement, genome, flags=0)]\n",
    "    REF = ['C']*len(targets_coordinates_TC) + ['G']*len(targets_coordinates_GA)\n",
    "    ALT = ['T']*len(targets_coordinates_TC) + ['A']*len(targets_coordinates_GA)\n",
    "    \n",
    "    data = {'#CHROM': [\"NC_063383.1\"]*len(REF),\n",
    "        'position': targets_coordinates_TC+targets_coordinates_GA, \n",
    "        \"REF\": REF,\n",
    "        \"ALT\": ALT}\n",
    "        \n",
    "    df = pd.DataFrame(data)\n",
    "    return df\n",
    "\n",
    "\n",
    "\n",
    "## find amino acids\n",
    "def load_genome(fasta_file):\n",
    "    \"\"\"Load the genome from a FASTA file.\"\"\"\n",
    "    genome = SeqIO.to_dict(SeqIO.parse(fasta_file, \"fasta\"))\n",
    "    return genome\n",
    "\n",
    "def load_gtf(gtf_file):\n",
    "    \"\"\"Load the GTF file into a pandas DataFrame.\"\"\"\n",
    "    gtf_columns = [\"seqname\", \"source\", \"feature\", \"start\", \"end\", \"score\", \"strand\", \"frame\", \"attribute\"]\n",
    "    gtf = pd.read_csv(gtf_file, sep=\"\\t\", comment='#', header=None, names=gtf_columns)\n",
    "    return gtf\n",
    "\n",
    "def get_cds(gtf, chromosome, position):\n",
    "    \"\"\"Find the CDS where the mutation is located.\"\"\"\n",
    "    cds_rows = gtf[(gtf['seqname'] == chromosome) & \n",
    "                   (gtf['feature'] == 'CDS') & \n",
    "                   (gtf['start'] <= position) & \n",
    "                   (gtf['end'] >= position)]\n",
    "    return cds_rows\n",
    "\n",
    "def translate_codon(codon):\n",
    "    \"\"\"Translate a codon to its corresponding amino acid.\"\"\"\n",
    "    return str(Seq(codon).translate(table=CodonTable.unambiguous_dna_by_id[11]))\n",
    "\n",
    "def detect_mutation_impact(genome, gtf, chromosome, position, ref_allele, alt_allele):\n",
    "    \"\"\"Detect the amino acid change caused by a mutation.\"\"\"\n",
    "    cds = get_cds(gtf, chromosome, position)\n",
    "    \n",
    "    if cds.empty:\n",
    "        return \"No CDS found at this position.\"\n",
    "    \n",
    "    strand = cds.iloc[0]['strand']\n",
    "        \n",
    "    gene_start = cds.iloc[0]['start']\n",
    "    \n",
    "    # Calculate the codon position\n",
    "    codon_start = gene_start + ((position - gene_start) // 3) * 3\n",
    "    \n",
    "    # Extract the codon\n",
    "    codon_seq = genome[chromosome].seq[codon_start-1:codon_start+2]\n",
    "    if strand == '-':\n",
    "        codon_seq = codon_seq.reverse_complement()\n",
    "        alt_allele = str(Seq(alt_allele).reverse_complement())\n",
    "    \n",
    "    # Translate the original codon\n",
    "    original_amino_acid = translate_codon(codon_seq)\n",
    "    \n",
    "    # Create the mutated codon\n",
    "    mutation_index = (position - codon_start) % 3\n",
    "    if strand == '-':\n",
    "        if mutation_index == 0:\n",
    "            mutation_index = 2\n",
    "        elif mutation_index == 2:\n",
    "            mutation_index = 0\n",
    "\n",
    "    mutated_codon = list(codon_seq)\n",
    "    mutated_codon[mutation_index] = alt_allele\n",
    "    mutated_codon = Seq(''.join(mutated_codon))\n",
    "    \n",
    "    # Translate the mutated codon\n",
    "    mutated_amino_acid = translate_codon(mutated_codon)\n",
    "    \n",
    "    \n",
    "    return original_amino_acid, mutated_amino_acid, codon_seq, mutated_codon, mutation_index+1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "c3ae8ae8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>position</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>aa_position</th>\n",
       "      <th>parent_codon</th>\n",
       "      <th>parent_aa</th>\n",
       "      <th>child_codon</th>\n",
       "      <th>child_aa</th>\n",
       "      <th>mutation_category</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_063383.1</td>\n",
       "      <td>23</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>intergenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_063383.1</td>\n",
       "      <td>60</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>intergenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_063383.1</td>\n",
       "      <td>78</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>intergenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_063383.1</td>\n",
       "      <td>82</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>intergenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_063383.1</td>\n",
       "      <td>98</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>intergenic</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        #CHROM  position REF ALT aa_position parent_codon parent_aa  \\\n",
       "0  NC_063383.1        23   C   T           -            -         -   \n",
       "1  NC_063383.1        60   C   T           -            -         -   \n",
       "2  NC_063383.1        78   C   T           -            -         -   \n",
       "3  NC_063383.1        82   C   T           -            -         -   \n",
       "4  NC_063383.1        98   C   T           -            -         -   \n",
       "\n",
       "  child_codon child_aa mutation_category  \n",
       "0           -        -        intergenic  \n",
       "1           -        -        intergenic  \n",
       "2           -        -        intergenic  \n",
       "3           -        -        intergenic  \n",
       "4           -        -        intergenic  "
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## create dataframe with potential APOBEC targets\n",
    "df_targets = find_potential_targets()\n",
    "\n",
    "## find amino acids\n",
    "genome = load_genome(\"data/MPOX_genome/NC_063383.1.fasta\")\n",
    "gtf = load_gtf(\"data/MPOX_genome/GCF_014621545.1_ASM1462154v1_genomic.220824.gff\")\n",
    "chromosome = \"NC_063383.1\"\n",
    "\n",
    "\n",
    "original_aa_list = []\n",
    "mutated_aa_list = []\n",
    "original_codon_list = []\n",
    "mutated_codon_list = []\n",
    "nucl_pos_list = []\n",
    "mutation_category = []\n",
    "for row in range(0, len(df_targets)):\n",
    "    position = df_targets.loc[row, \"position\"]\n",
    "    ref_allele = df_targets.loc[row, \"REF\"]\n",
    "    alt_allele = df_targets.loc[row, \"ALT\"]\n",
    "    \n",
    "    try:\n",
    "        original_aa, mutated_aa, original_codon, mutated_codon, nucl_pos = detect_mutation_impact(genome, gtf, chromosome, position, ref_allele, alt_allele)\n",
    "    except:\n",
    "        mutation_category.append(\"intergenic\")\n",
    "        original_aa_list.append('-')\n",
    "        mutated_aa_list.append('-')\n",
    "        original_codon_list.append('-')\n",
    "        mutated_codon_list.append('-')\n",
    "        nucl_pos_list.append('-')\n",
    "    else:\n",
    "        original_aa_list.append(original_aa)\n",
    "        mutated_aa_list.append(mutated_aa)\n",
    "        original_codon_list.append(original_codon)\n",
    "        mutated_codon_list.append(mutated_codon)\n",
    "        nucl_pos_list.append(nucl_pos)\n",
    "        if mutated_aa == original_aa:\n",
    "            mutation_category.append(\"synonymous\")\n",
    "        elif mutated_aa == \"*\":\n",
    "            mutation_category.append(\"nonsense\")\n",
    "        else:\n",
    "            mutation_category.append(\"nonsynonymous\")\n",
    "\n",
    "\n",
    "df_targets[\"aa_position\"] = nucl_pos_list\n",
    "df_targets[\"parent_codon\"] = original_codon_list\n",
    "df_targets[\"parent_aa\"] = original_aa_list\n",
    "df_targets[\"child_codon\"] = mutated_codon_list\n",
    "df_targets[\"child_aa\"] = mutated_aa_list\n",
    "df_targets[\"mutation_category\"] = mutation_category\n",
    "df_targets.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d5e11405",
   "metadata": {},
   "source": [
    "Saving tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "ac3a6525",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_targets.to_csv(\"data/APOBEC_targets_aa.csv\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "12a48591",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>site</th>\n",
       "      <th>REF_v2</th>\n",
       "      <th>ALT</th>\n",
       "      <th>aa_position_FIXED</th>\n",
       "      <th>parent_codon_FIXED</th>\n",
       "      <th>parent_aa_FIXED</th>\n",
       "      <th>child_codon_FIXED</th>\n",
       "      <th>child_aa_FIXED</th>\n",
       "      <th>mutation_category_FIXED</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_063383.1</td>\n",
       "      <td>186165</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>GAT</td>\n",
       "      <td>D</td>\n",
       "      <td>AAT</td>\n",
       "      <td>N</td>\n",
       "      <td>nonsynonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_063383.1</td>\n",
       "      <td>55133</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>CGT</td>\n",
       "      <td>R</td>\n",
       "      <td>TGT</td>\n",
       "      <td>C</td>\n",
       "      <td>nonsynonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_063383.1</td>\n",
       "      <td>55133</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>CGT</td>\n",
       "      <td>R</td>\n",
       "      <td>TGT</td>\n",
       "      <td>C</td>\n",
       "      <td>nonsynonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_063383.1</td>\n",
       "      <td>55133</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>CGT</td>\n",
       "      <td>R</td>\n",
       "      <td>TGT</td>\n",
       "      <td>C</td>\n",
       "      <td>nonsynonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_063383.1</td>\n",
       "      <td>55133</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>3</td>\n",
       "      <td>CTC</td>\n",
       "      <td>L</td>\n",
       "      <td>CTT</td>\n",
       "      <td>L</td>\n",
       "      <td>synonymous</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        #CHROM    site REF_v2 ALT aa_position_FIXED parent_codon_FIXED  \\\n",
       "0  NC_063383.1  186165      G   A                 1                GAT   \n",
       "1  NC_063383.1   55133      G   A                 1                CGT   \n",
       "2  NC_063383.1   55133      G   A                 1                CGT   \n",
       "3  NC_063383.1   55133      G   A                 1                CGT   \n",
       "4  NC_063383.1   55133      G   A                 3                CTC   \n",
       "\n",
       "  parent_aa_FIXED child_codon_FIXED child_aa_FIXED mutation_category_FIXED  \n",
       "0               D               AAT              N           nonsynonymous  \n",
       "1               R               TGT              C           nonsynonymous  \n",
       "2               R               TGT              C           nonsynonymous  \n",
       "3               R               TGT              C           nonsynonymous  \n",
       "4               L               CTT              L              synonymous  "
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1_new['#CHROM'] = \"NC_063383.1\"\n",
    "df1_new = df1_new[['#CHROM', 'site', 'REF_v2', 'ALT', 'aa_position_FIXED', 'parent_codon_FIXED', 'parent_aa_FIXED','child_codon_FIXED','child_aa_FIXED','mutation_category_FIXED']]\n",
    "df1_new.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "a2f29db7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>position</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>aa_position</th>\n",
       "      <th>parent_codon</th>\n",
       "      <th>parent_aa</th>\n",
       "      <th>child_codon</th>\n",
       "      <th>child_aa</th>\n",
       "      <th>mutation_category</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_063383.1</td>\n",
       "      <td>186165</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>GAT</td>\n",
       "      <td>D</td>\n",
       "      <td>AAT</td>\n",
       "      <td>N</td>\n",
       "      <td>nonsynonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_063383.1</td>\n",
       "      <td>55133</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>CGT</td>\n",
       "      <td>R</td>\n",
       "      <td>TGT</td>\n",
       "      <td>C</td>\n",
       "      <td>nonsynonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_063383.1</td>\n",
       "      <td>55133</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>CGT</td>\n",
       "      <td>R</td>\n",
       "      <td>TGT</td>\n",
       "      <td>C</td>\n",
       "      <td>nonsynonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_063383.1</td>\n",
       "      <td>55133</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>CGT</td>\n",
       "      <td>R</td>\n",
       "      <td>TGT</td>\n",
       "      <td>C</td>\n",
       "      <td>nonsynonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_063383.1</td>\n",
       "      <td>55133</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>3</td>\n",
       "      <td>CTC</td>\n",
       "      <td>L</td>\n",
       "      <td>CTT</td>\n",
       "      <td>L</td>\n",
       "      <td>synonymous</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        #CHROM  position REF ALT aa_position parent_codon parent_aa  \\\n",
       "0  NC_063383.1    186165   G   A           1          GAT         D   \n",
       "1  NC_063383.1     55133   G   A           1          CGT         R   \n",
       "2  NC_063383.1     55133   G   A           1          CGT         R   \n",
       "3  NC_063383.1     55133   G   A           1          CGT         R   \n",
       "4  NC_063383.1     55133   G   A           3          CTC         L   \n",
       "\n",
       "  child_codon child_aa mutation_category  \n",
       "0         AAT        N     nonsynonymous  \n",
       "1         TGT        C     nonsynonymous  \n",
       "2         TGT        C     nonsynonymous  \n",
       "3         TGT        C     nonsynonymous  \n",
       "4         CTT        L        synonymous  "
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1_new.columns = df_targets.columns\n",
    "df1_new.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "c81e67c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "df1_new.to_csv(\"data/B1_data_fixed_APOBEC.csv\", sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "36e5f58f",
   "metadata": {},
   "source": [
    "Amino acids hypothetical changes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "db105448",
   "metadata": {},
   "outputs": [],
   "source": [
    "from dash import Dash, dcc, html, Input, Output\n",
    "import plotly.graph_objects as go\n",
    "import json, urllib\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "3bd78216",
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: 'data/grantham_score_B1.csv'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[41], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m df \u001b[38;5;241m=\u001b[39m \u001b[43mpd\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread_csv\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mdata/grantham_score_B1.csv\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msep\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m,\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m      2\u001b[0m df \u001b[38;5;241m=\u001b[39m df\u001b[38;5;241m.\u001b[39mreset_index()\n\u001b[1;32m      3\u001b[0m df[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mparent_aa_FIXED\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m [x\u001b[38;5;241m+\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m_observed\u001b[39m\u001b[38;5;124m'\u001b[39m \u001b[38;5;28;01mfor\u001b[39;00m x \u001b[38;5;129;01min\u001b[39;00m df\u001b[38;5;241m.\u001b[39mparent_aa\u001b[38;5;241m.\u001b[39mtolist()]\n",
      "File \u001b[0;32m/opt/homebrew/lib/python3.12/site-packages/pandas/util/_decorators.py:211\u001b[0m, in \u001b[0;36mdeprecate_kwarg.<locals>._deprecate_kwarg.<locals>.wrapper\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    209\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    210\u001b[0m         kwargs[new_arg_name] \u001b[38;5;241m=\u001b[39m new_arg_value\n\u001b[0;32m--> 211\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfunc\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/opt/homebrew/lib/python3.12/site-packages/pandas/util/_decorators.py:331\u001b[0m, in \u001b[0;36mdeprecate_nonkeyword_arguments.<locals>.decorate.<locals>.wrapper\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    325\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(args) \u001b[38;5;241m>\u001b[39m num_allow_args:\n\u001b[1;32m    326\u001b[0m     warnings\u001b[38;5;241m.\u001b[39mwarn(\n\u001b[1;32m    327\u001b[0m         msg\u001b[38;5;241m.\u001b[39mformat(arguments\u001b[38;5;241m=\u001b[39m_format_argument_list(allow_args)),\n\u001b[1;32m    328\u001b[0m         \u001b[38;5;167;01mFutureWarning\u001b[39;00m,\n\u001b[1;32m    329\u001b[0m         stacklevel\u001b[38;5;241m=\u001b[39mfind_stack_level(),\n\u001b[1;32m    330\u001b[0m     )\n\u001b[0;32m--> 331\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfunc\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/opt/homebrew/lib/python3.12/site-packages/pandas/io/parsers/readers.py:950\u001b[0m, in \u001b[0;36mread_csv\u001b[0;34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, squeeze, prefix, mangle_dupe_cols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, encoding_errors, dialect, error_bad_lines, warn_bad_lines, on_bad_lines, delim_whitespace, low_memory, memory_map, float_precision, storage_options)\u001b[0m\n\u001b[1;32m    935\u001b[0m kwds_defaults \u001b[38;5;241m=\u001b[39m _refine_defaults_read(\n\u001b[1;32m    936\u001b[0m     dialect,\n\u001b[1;32m    937\u001b[0m     delimiter,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    946\u001b[0m     defaults\u001b[38;5;241m=\u001b[39m{\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdelimiter\u001b[39m\u001b[38;5;124m\"\u001b[39m: \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m,\u001b[39m\u001b[38;5;124m\"\u001b[39m},\n\u001b[1;32m    947\u001b[0m )\n\u001b[1;32m    948\u001b[0m kwds\u001b[38;5;241m.\u001b[39mupdate(kwds_defaults)\n\u001b[0;32m--> 950\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_read\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/opt/homebrew/lib/python3.12/site-packages/pandas/io/parsers/readers.py:605\u001b[0m, in \u001b[0;36m_read\u001b[0;34m(filepath_or_buffer, kwds)\u001b[0m\n\u001b[1;32m    602\u001b[0m _validate_names(kwds\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mnames\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m))\n\u001b[1;32m    604\u001b[0m \u001b[38;5;66;03m# Create the parser.\u001b[39;00m\n\u001b[0;32m--> 605\u001b[0m parser \u001b[38;5;241m=\u001b[39m \u001b[43mTextFileReader\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    607\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m chunksize \u001b[38;5;129;01mor\u001b[39;00m iterator:\n\u001b[1;32m    608\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m parser\n",
      "File \u001b[0;32m/opt/homebrew/lib/python3.12/site-packages/pandas/io/parsers/readers.py:1442\u001b[0m, in \u001b[0;36mTextFileReader.__init__\u001b[0;34m(self, f, engine, **kwds)\u001b[0m\n\u001b[1;32m   1439\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39moptions[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhas_index_names\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m kwds[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhas_index_names\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n\u001b[1;32m   1441\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles: IOHandles \u001b[38;5;241m|\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[0;32m-> 1442\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_engine \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_make_engine\u001b[49m\u001b[43m(\u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mengine\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/opt/homebrew/lib/python3.12/site-packages/pandas/io/parsers/readers.py:1735\u001b[0m, in \u001b[0;36mTextFileReader._make_engine\u001b[0;34m(self, f, engine)\u001b[0m\n\u001b[1;32m   1733\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m mode:\n\u001b[1;32m   1734\u001b[0m         mode \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m-> 1735\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles \u001b[38;5;241m=\u001b[39m \u001b[43mget_handle\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1736\u001b[0m \u001b[43m    \u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1737\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmode\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1738\u001b[0m \u001b[43m    \u001b[49m\u001b[43mencoding\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mencoding\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1739\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcompression\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mcompression\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1740\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmemory_map\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mmemory_map\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1741\u001b[0m \u001b[43m    \u001b[49m\u001b[43mis_text\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mis_text\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1742\u001b[0m \u001b[43m    \u001b[49m\u001b[43merrors\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mencoding_errors\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mstrict\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1743\u001b[0m \u001b[43m    \u001b[49m\u001b[43mstorage_options\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mstorage_options\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1744\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1745\u001b[0m \u001b[38;5;28;01massert\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m   1746\u001b[0m f \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles\u001b[38;5;241m.\u001b[39mhandle\n",
      "File \u001b[0;32m/opt/homebrew/lib/python3.12/site-packages/pandas/io/common.py:856\u001b[0m, in \u001b[0;36mget_handle\u001b[0;34m(path_or_buf, mode, encoding, compression, memory_map, is_text, errors, storage_options)\u001b[0m\n\u001b[1;32m    851\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(handle, \u001b[38;5;28mstr\u001b[39m):\n\u001b[1;32m    852\u001b[0m     \u001b[38;5;66;03m# Check whether the filename is to be opened in binary mode.\u001b[39;00m\n\u001b[1;32m    853\u001b[0m     \u001b[38;5;66;03m# Binary mode does not support 'encoding' and 'newline'.\u001b[39;00m\n\u001b[1;32m    854\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m ioargs\u001b[38;5;241m.\u001b[39mencoding \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m ioargs\u001b[38;5;241m.\u001b[39mmode:\n\u001b[1;32m    855\u001b[0m         \u001b[38;5;66;03m# Encoding\u001b[39;00m\n\u001b[0;32m--> 856\u001b[0m         handle \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mopen\u001b[39;49m\u001b[43m(\u001b[49m\n\u001b[1;32m    857\u001b[0m \u001b[43m            \u001b[49m\u001b[43mhandle\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    858\u001b[0m \u001b[43m            \u001b[49m\u001b[43mioargs\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmode\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    859\u001b[0m \u001b[43m            \u001b[49m\u001b[43mencoding\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mioargs\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mencoding\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    860\u001b[0m \u001b[43m            \u001b[49m\u001b[43merrors\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43merrors\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    861\u001b[0m \u001b[43m            \u001b[49m\u001b[43mnewline\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    862\u001b[0m \u001b[43m        \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    863\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    864\u001b[0m         \u001b[38;5;66;03m# Binary mode\u001b[39;00m\n\u001b[1;32m    865\u001b[0m         handle \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mopen\u001b[39m(handle, ioargs\u001b[38;5;241m.\u001b[39mmode)\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: 'data/grantham_score_B1.csv'"
     ]
    }
   ],
   "source": [
    "df = pd.read_csv(\"data/grantham_score_B1_APOBEC.csv\", sep=',')\n",
    "df = df.reset_index()\n",
    "df['parent_aa_FIXED'] = [x+'_observed' for x in df.parent_aa.tolist()]\n",
    "df['child_aa_FIXED'] = [x+'_editing' for x in df.child_aa.tolist()]\n",
    "df1 = df.groupby(['parent_aa_FIXED','child_aa_FIXED', 'grantham_rank_color'])['index'].count().reset_index()\n",
    "df1.columns = ['source','target', 'grantham_rank_color', 'value']\n",
    "df1 = df1.sort_values(by='value', ascending=True)\n",
    "\n",
    "unique_source_target = list(pd.unique(df1[['source','target']].values.ravel('k')))\n",
    "\n",
    "unique_source_target_label = [list(x)[0] for x in unique_source_target]\n",
    "\n",
    "mapping_dict = {k: v for v, k in enumerate(unique_source_target)}\n",
    "#print(mapping_dict)\n",
    "\n",
    "df1['source'] = df1['source'].map(mapping_dict)\n",
    "df1['target'] = df1['target'].map(mapping_dict)\n",
    "#print(df1)\n",
    "\n",
    "new_dict = df1.to_dict(orient='list')\n",
    "#print(new_dict)\n",
    "\n",
    "\n",
    "fig = go.Figure(data=[go.Sankey(\n",
    "    node = dict(\n",
    "        pad = 20,\n",
    "        thickness=30,\n",
    "        line=dict(color='black', width=0.5),\n",
    "        label = unique_source_target_label,\n",
    "        color='#81B29A'\n",
    "    ),\n",
    "    link = dict(\n",
    "        source= new_dict['source'],\n",
    "        target = new_dict['target'],\n",
    "        value = new_dict['value'],\n",
    "        color = new_dict['grantham_rank_color']\n",
    "    )\n",
    "\n",
    ")\n",
    "])\n",
    "#legend={\"yellow\": \"conservative\", \"pink\": \"moderately conservative\", \"dark blue\": \"moderately radical\", \"brown\": \"radical\", \"green\": \"synonymous\"}\n",
    "fig.update_layout(\n",
    "    font_family=\"Arial\",\n",
    "    font_color=\"black\",\n",
    "    font_size=18,\n",
    "    width=700,\n",
    "    height=1000\n",
    ")\n",
    "\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "3ed6f26d",
   "metadata": {},
   "outputs": [],
   "source": [
    "outfile = \"data/aa_hypothetical_changes.html\"\n",
    "fig.write_html(outfile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "66529ce1",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
